[DOCS] Updating Weight Compression Article (#27432)
openvinotoolkit/openvino: docs/articles_en/openvino-workflow/model-optimization-guide/weight-compression.rst

   :hidden:

   weight-compression/microscaling-quantization
   weight-compression/4-bit-weight-quantization

Weight compression enhances the efficiency of models,
which have large memory requirements. This method reduces the model's
memory footprint, a crucial factor for Large Language Models (LLMs).

> Unlike full model quantization, where weights and activations are quantized,
> weight compression in `Neural Network Compression Framework (NNCF) <https://github.com/openvinotoolkit/nncf>`__

Reviewer: ```suggestion
Unlike full model quantization, where both weights and activations are quantized, it
```

only targets the model's weights. This approach enables the activations to remain as
floating-point numbers, preserving most of the model's accuracy while improving its
speed and reducing its size. The reduction in size is especially noticeable with larger models.
For instance the 7 billion parameter Llama 2 model can be reduced
from about 25GB to 4GB using 4-bit weight compression.

.. note::

   With smaller language models (i.e. less than 1B parameters), weight
   compression may result in more accuracy reduction than with larger models.
   Therefore, weight compression is recommended for use with LLMs only.

LLMs and other models that require
extensive memory to store the weights during inference can benefit

> from weight compression as it:

Reviewer: Maybe we don't need this piece. It just restates stuff. But we need to check what "other models" means in this case.

Author: > Maybe we don't need this piece. It just restates stuff. But we need to check what "other models" means in this case.

I have changed it with the [commit](https://github.com/openvinotoolkit/openvino/pull/27432/commits/b354ec794b5d3ef1ef25ae26a61cddfa9ce6dff5)
as agreed with @AlexKoff88

method primarily designed to optimize LLMs.

Compression Methods (8-bit vs. 4-bit)
#####################################

> **8-bit weight quantization** method offers a balance between model size reduction and
> maintaining accuracy, which usually leads to significant performance improvements for
> Transformer-based models. Models with 8-bit compressed weights are performant on the
> vast majority of supported CPU and GPU platforms. By default, weights are compressed
> asymmetrically to "INT8_ASYM" mode.

Reviewer: ```suggestion
**8-bit weight quantization** offers a good balance between reducing the size and lowering the
accuracy of a model. It usually results in significant improvements for transformer-based models
and guarantees good model performance for a vast majority of supported CPU and GPU platforms.
By default, weights are compressed asymmetrically to "INT8_ASYM" mode.
```

> The code snippet below shows how to do asymmetrical 8-bit quantization of the model weights
> represented in OpenVINO IR using NNCF:

Reviewer: ```suggestion
Here is an example of code using NNCF to perform asymmetrical 8-bit weight quantization of
a model in the OpenVINO IR format:
```

        model = OVModelForCausalLM.from_pretrained(model_id, export=True, load_in_8bit=False, compile=False)
        model.model = compress_weights(model.model, mode=CompressWeightsMode.INT8_ASYM)

        # Inference
        model.compile()
        tokenizer = AutoTokenizer.from_pretrained(model_id)
        pipe = pipeline("text-generation", model=model, tokenizer=tokenizer)
        phrase = "The weather is"
        results = pipe(phrase)
        print(results)

   .. tab-item:: Compression with Optimum-Intel
      :sync: optimum

      Load a pre-trained Hugging Face model, compress it to INT8_ASYM, using the
      Optimum Intel API, and then execute inference with a text phrase:

      .. code-block:: python

         from optimum.intel.openvino import OVModelForCausalLM, OVWeightQuantizationConfig
         from transformers import AutoTokenizer, pipeline

         # Load and compress a model from Hugging Face.
         model_id = "microsoft/Phi-3.5-mini-instruct"
         model = OVModelForCausalLM.from_pretrained(
             model_id,
             export=True,
             quantization_config=OVWeightQuantizationConfig(
                 bits=8,

>                  quant_method="awq",

Reviewer 2: This is a config for int4 quantization (not int8) and it should be moved below with the change of `bits=4` and the comment above.
Also you have the command-line variant of it which is for int4 as you can see `optimum-cli export openvino --model microsoft/Phi-3.5-mini-instruct --weight-format int4 --awq --scale-estimation --dataset wikitext2 --group-size 64 --ratio 1.0 ov_phi-3.5-mini-instruct`.

For INT8 it looks like:
```
quantization_config=OVWeightQuantizationConfig(bits=8)
```
and
```
optimum-cli export openvino --model microsoft/Phi-3.5-mini-instruct --weight-format int8 ov_phi-3.5-mini-instruct
```

Author: I have applied necessary changes in: https://github.com/openvinotoolkit/openvino/pull/27432/commits/5b49f023635c0ac453d5ead14e57f4fc149407c3

        model.model = compress_weights(model.model, mode=CompressWeightsMode.INT8_ASYM)

        # Inference
        model.compile()
        tokenizer = AutoTokenizer.from_pretrained(model_id)
        pipe = pipeline("text-generation", model=model, tokenizer=tokenizer)
        phrase = "The weather is"
        results = pipe(phrase)
        print(results)

   .. tab-item:: Compression with Optimum-Intel
      :sync: optimum

      Load a pre-trained Hugging Face model, compress it to INT8_ASYM, using the
      Optimum Intel API, and then execute inference with a text phrase:

      .. code-block:: python

         from optimum.intel.openvino import OVModelForCausalLM
         from transformers import AutoTokenizer, pipeline

         # Load a model and compress it with Optimum Intel.
         model_id = "HuggingFaceH4/zephyr-7b-beta"
         model = OVModelForCausalLM.from_pretrained(model_id, export=True, load_in_8bit=True)

         # Inference
         tokenizer = AutoTokenizer.from_pretrained(model_id)
         pipe = pipeline("text-generation", model=model, tokenizer=tokenizer)
         phrase = "The weather is"
         results = pipe(phrase)
         print(results)

> The model can be also :ref:`saved into a compressed format <save_pretrained>`,

Reviewer: ```suggestion
The model can also be :ref:`saved into a compressed format <save_pretrained>`,
```

resulting in a smaller binary file.

> **4-bit weight quantization** method stands for an INT4-INT8 mixed-precision weight quantization,
> where INT4 is considered as the primary precision and asymmetric INT8 is the backup one.
> It usually results in a smaller model size and lower inference latency, although the accuracy

Reviewer: ```suggestion
**4-bit weight quantization** is actually a mixed-precision compression, primarily INT4 and
a backup asymmetric INT8 precisions. It produces a smaller model, offering lower inference latency but potentially noticeable accuracy degradation, depending on the model.
```

> degradation can be higher, depending on the model.

Reviewer: ```suggestion

```
